t/fixtures/proxy.lua

```diff
@@ -0,0 +1,222 @@
+
+        if header and str_lower(header.name) == 'content-length' then
+            body_length = tonumber(header.value)
+            body_reader = _forward_body
+        end
+
+        if header and str_lower(header.name) == 'transfer-encoding' and str_lower(header.value) == 'chunked' then
+            body_reader = _forward_chunked_body
+        end
+    until header_line == cr_lf
+
+    body_reader(sock, upstream, body_length)
+end
+
+local function proxy_http_request(sock, method, url, http_version, upstream)
+    local uri = assert(http.parse_uri(nil, url))
+    local port = uri[3]
+    local host = uri[2]
+    local ok, err = upstream:connect(host, port)
+
+    if ok then
+        ngx.log(ngx.DEBUG, 'connected to upstream ', host, ':', port)
+        ngx.ctx.upstream = upstream
+        ngx.ctx.upstream_host = host
+        ngx.ctx.upstream_port = port
+    else
+        ngx.log(ngx.ERR, 'failed to connect to ', host, ':', port, ' err: ', err)
+        return nil, err
+    end
+
+    local req_line = format('%s %s %s%s', method, uri[4] or '/', http_version, cr_lf)
+
+    send(upstream, req_line)
+
+    forward_http_stream(sock, upstream)
+    forward_http_stream(upstream, sock)
+    ngx.log(ngx.DEBUG, 'finished forwarding http stream')
+
+    upstream:setkeepalive()
+end
+
+local function proxy_connect_request(sock, connect, upstream)
+    local ok, err = upstream:connect(connect.host, connect.port)
+
+    if ok then
+        ngx.log(ngx.DEBUG, 'connected to upstream ', connect.host, ':', connect.port)
+        ngx.ctx.upstream = upstream
+        ngx.ctx.upstream_host = connect.host
+        ngx.ctx.upstream_port = connect.port
+    else
+        ngx.log(ngx.ERR, 'failed to connect to ', connect.host, ':', connect.port, ' err: ', err)
+        return nil, err
+    end
+
+    local header_line
+    local read_request_header = sock:receiveuntil(cr_lf, inclusive)
+
+    repeat
+        header_line = read_request_header()
+        ngx.log(ngx.DEBUG, 'got header line: ', header_line)
+    until header_line == cr_lf
+
+    send(sock, http_ok)
+    send(sock, cr_lf)
+
+    return pipe_stream(sock, upstream)
+end
+
+return function()
+    local client, err = ngx.req.socket(true)
+
+    if not client then return nil, err end
+
+    local readline = client:receiveuntil(cr_lf)
+
+    while true do
+        local data, partial
+        data, err, partial = readline()
+
+        if data then
+            ngx.log(ngx.DEBUG, 'got line: ', data)
+        else
+            ngx.log(ngx.WARN, 'got partial: ', partial, ' err: ', err)
+            ngx.exit(ngx.OK)
+        end
+
+        local match = ngx_re_match(data, [[^(?<method>[A-Z]+)\s(?<uri>\S+)\s(?<http>\S+)$]], 'oj')
+
+        if not match then
+            ngx.log(ngx.WARN, 'got invalid request: ', data)
+            return client:close()
+        end
+
+        local upstream = ngx.socket.tcp()
+
+        ngx.log(ngx.DEBUG, 'proxy request: ', data)
+        ngx.ctx.request = data
```

Comment:
Why do we need to store this in the context?

Reply:
Will describe it in the comment, but basically so we can print `CONNECT host:port` later when reusing keepalive connection so the test can match that.